pkgs/ffigen/lib/src/code_generator/objc_methods.dart

```diff
@@ -0,0 +1,224 @@
+    // override it.
+    if (newMethod.isProperty && !oldMethod.isProperty) {
+      return newMethod;
+    } else if (!newMethod.isProperty && oldMethod.isProperty) {
+      // Don't override, but also skip the same method check below.
+      return oldMethod;
+    }
+
+    // Check the duplicate is the same method.
+    if (!newMethod.sameAs(oldMethod)) {
+      _logger.severe('Duplicate methods with different signatures: '
+          '$originalName.${newMethod.originalName}');
+      return newMethod;
+    }
+
+    // There's a bug in some Apple APIs where an init method that should return
+    // instancetype has a duplicate definition that instead returns id. In that
+    // case, use the one that returns instancetype. Note that since instancetype
+    // is an alias of id, the sameAs check above passes.
+    if (ObjCBuiltInFunctions.isInstanceType(newMethod.returnType) &&
+        !ObjCBuiltInFunctions.isInstanceType(oldMethod.returnType)) {
+      return newMethod;
+    } else if (!ObjCBuiltInFunctions.isInstanceType(newMethod.returnType) &&
+        ObjCBuiltInFunctions.isInstanceType(oldMethod.returnType)) {
+      return oldMethod;
+    }
+
+    return newMethod;
+  }
+
+  bool _shouldIncludeMethod(ObjCMethod method) =>
+      method.childTypes.every((Type t) {
+        t = t.typealiasType.baseType;
+
+        // Ignore methods with variadic args.
+        // TODO(https://github.com/dart-lang/native/issues/1192): Remove this.
+        if (t is Struct && t.originalName == '__va_list_tag') {
+          return false;
+        }
+
+        // Ignore methods with block args or rets when we're generating in
+        // package:objective_c.
+        // TODO(https://github.com/dart-lang/native/issues/1180): Remove this.
+        if (builtInFunctions.generateForPackageObjectiveC && t is ObjCBlock) {
+          return false;
+        }
+
+        return true;
+      });
+}
+
+enum ObjCMethodKind {
+  method,
+  propertyGetter,
+  propertySetter,
+}
+
+class ObjCProperty {
+  final String originalName;
+  String? dartName;
+
+  ObjCProperty(this.originalName);
+}
+
+class ObjCMethod {
+  final String? dartDoc;
+  final String originalName;
+  final ObjCProperty? property;
+  Type returnType;
+  final List<ObjCMethodParam> params;
+  final ObjCMethodKind kind;
+  final bool isClass;
+  final bool isOptional;
+  bool returnsRetained = false;
+  ObjCInternalGlobal? selObject;
+  ObjCMsgSendFunc? msgSend;
+  ObjCBlock? protocolBlock;
+
+  ObjCMethod({
+    required this.originalName,
+    this.property,
+    this.dartDoc,
+    required this.kind,
+    required this.isClass,
+    required this.isOptional,
+    required this.returnType,
+    List<ObjCMethodParam>? params_,
+  }) : params = params_ ?? [];
+
+  bool get isProperty =>
+      kind == ObjCMethodKind.propertyGetter ||
+      kind == ObjCMethodKind.propertySetter;
+  bool get isRequired => !isOptional;
+  bool get isInstance => !isClass;
```

Comment:
isClass and isInstance could do with documentation that it is an instance method and class method. (A method is never a class itself.) Renaming to something longer could also work isClassMethod isInstanceMethod (but more verbose).

Reply:
Done